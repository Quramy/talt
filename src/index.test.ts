import ts from "typescript";

import { template, printNode } from "./";

describe("Node type", () => {
  test(template.typeNode.name, () => {
    const node = template.typeNode<ts.TypeLiteralNode>`{ a: 1 }`();
    expect(ts.isTypeLiteralNode(node)).toBeTruthy();
    expect(printNode(node)).toMatchSnapshot();
  });

  test(template.expression.name, () => {
    const node = template.expression<ts.ObjectLiteralExpression>`{ a: 1 }`();
    expect(ts.isObjectLiteralExpression(node)).toBeTruthy();
    expect(printNode(node)).toMatchSnapshot();
  });

  test(template.statement.name, () => {
    const node = template.statement<ts.TypeAliasDeclaration>`type a = 100`();
    expect(ts.isTypeAliasDeclaration(node)).toBeTruthy();
    expect(printNode(node)).toMatchSnapshot();
  });

  test(template.jsxAttribute.name, () => {
    const node = template.jsxAttribute<ts.JsxAttribute>`id={id}`();
    expect(ts.isJsxAttribute(node)).toBeTruthy();
    expect(printNode(node)).toMatchSnapshot();
  });

  test(template.sourceFile.name, () => {
    const node = template.sourceFile`type a = 100`();
    expect(ts.isSourceFile(node)).toBeTruthy();
    expect(printNode(node)).toMatchSnapshot();
  });
});

describe("Replacement", () => {
  test("compiled function generates new node instance", () => {
    const fn = template.expression("100 + 100");
    const nodeA = fn();
    const nodeB = fn();
    expect(nodeA === nodeB).toBeFalsy();
  });

  test("string placeholder", () => {
    const idA = "A";
    const idB = "B";
    const node = template.typeNode`{ a: ${idA}, b: ${idB} }`();
    expect(printNode(node)).toMatchSnapshot();
  });

  test("node bind", () => {
    const idA = ts.factory.createIdentifier("A");
    const idB = ts.factory.createIdentifier("B");
    const node = template.typeNode`{ a: ${idA}, b: ${idB} }`();
    expect(printNode(node)).toMatchSnapshot();
  });

  test("id placeholder", () => {
    const exp = template.expression`200 * 300`();
    const fn = template.expression`100 + TO_BE_REPLACED`;
    const node = fn({
      TO_BE_REPLACED: exp,
    });
    expect(printNode(node)).toMatchSnapshot();
  });
});
