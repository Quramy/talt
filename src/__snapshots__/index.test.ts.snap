// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node type expressionTag 1`] = `"{ a: 1 }"`;

exports[`Node type jsxAttributeTag 1`] = `"id={id}"`;

exports[`Node type sourceTag 1`] = `
"type a = 100;
"
`;

exports[`Node type statementTag 1`] = `"type a = 100;"`;

exports[`Node type typeTag 1`] = `
"{
    a: 1;
}"
`;

exports[`Replacement id placeholder anonymous function 1`] = `"100 * fuga * 10"`;

exports[`Replacement id placeholder identifier to identifier replacement at type reference 1`] = `"type X = After;"`;

exports[`Replacement id placeholder identifier to type node replacement at type reference 1`] = `"type X = {};"`;

exports[`Replacement id placeholder nested 1`] = `"100 * hoge * 200"`;

exports[`Replacement id placeholder replacement 1`] = `"100 + 200 * 300"`;

exports[`Replacement id placeholder same identifiers 1`] = `"100 + 200 * 300 + 200 * 300"`;

exports[`Replacement node bind 1`] = `
"{
    a: A;
    b: B;
}"
`;

exports[`Replacement string placeholder 1`] = `
"{
    a: A;
    b: B;
}"
`;
